test/app_deployment.yml.sh

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+. ./store.sh
+
+store_set app.CONJUR_AUTHN_URL "$(store_get CONJUR_URL)/authn-k8s/$(store_get AUTHENTICATOR_ID)"
+store_set app.CONJUR_APPLIANCE_URL "$(store_get CONJUR_URL)"
+store_set app.CONJUR_ACCOUNT "$(store_get CONJUR_ACCOUNT)"
+store_set app.CONJUR_AUTHN_LOGIN "host/conjur/authn-k8s/$(store_get AUTHENTICATOR_ID)/apps/$(store_get TEST_NAMESPACE)/service_account/$(store_get APP_SERVICE_ACCOUNT)"
+store_set app.CONJUR_SSL_CERTIFICATE_SECRET "$(store_get CONJUR_RELEASE_NAME)-conjur-ssl-cert"
+store_set app.APP_SERVICE_ACCOUNT "$(store_get APP_SERVICE_ACCOUNT)"
+store_set app.CONJUR_KUBERNETES_AUTHENTICATOR_RELEASE_VERSION "$(store_get CONJUR_KUBERNETES_AUTHENTICATOR_RELEASE_VERSION)"
+
+echo "
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: test-app
+  name: test-app
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: test-app
+  template:
+    metadata:
+      labels:
+        app: test-app
+    spec:
+      serviceAccountName: $(store_get app.APP_SERVICE_ACCOUNT)
```

Comment:
Instead of including the `store_get` executables in these self-executing scripts, could we instead:
* Write these templates as traditional YAML files, using the `{{ VARIABLE_NAME }}` syntax. So the above line would be:
```
      serviceAccountName: {{ APP_SERVICE_ACCOUNT }}
```
* Then create a translator that:
  - Scans the YAML file for the `{{ VARIABLE_NAME }}` pattern
  - Calls `store-get` for that variable name
  - Does a `sed` substitute for the value retrieved by `store-get`
If that's possible, the YAML would look much more traditional, and easier to translate from our existing YAML files.

Reply:
> Write these templates as traditional YAML files...

We had some places (demo scripts) where this was this case before. These would typically be stored as actual YAML files that would get substitution via `sed` as you suggested. I thought there were a few issues with this:
1. The template file was masquerade as YAML file which was a bit dishonest. Whereas `X.yml.sh` is quite explicit and only requires that the values being consumed be set.  
2. The template was using an ad-hoc templating syntax `{{ }}` so IDEs weren't offering much assistance
3. `sed` seemed limited in its ability to substitute arbitrary values, whereas having the full expressive capability of the shell meant anything was possible.

I think it'd definitely be a good idea to standardise on how values are passed to the template. Perhaps inlining `$(store_get app.APP_SERVICE_ACCOUNT)` is more verbose than need be. I was thinking we could perhaps fetch all the variable up top and in the YAML snippet we're only allowed to substitute environment variable e.g.
```
CONJUR_AUTHN_LOGIN="host/conjur/authn-k8s/$(store_get AUTHENTICATOR_ID)/apps/$(store_get TEST_NAMESPACE)/service_account/$(store_get APP_SERVICE_ACCOUNT)"

echo "
env:
- name: CONJUR_AUTHN_LOGIN
   value: '${CONJUR_AUTHN_LOGIN}'
"
```

Reply:
Personally, I have a slight preference for this inlining using env variables. It feels more intuitive doing direct env string substitution in the YAML template... or should I say "bash-script-yaml-template".

I'm sure we've looked into using YTT (YAML templating tool: https://get-ytt.io/)?  It looks like they use YAML comments for their templatized fields. I didn't look at this in depth, so I don't know if it has the richness/flexibility that we're looking for in our templates.